src/common/grpc/src/channel_manager.rs

```diff
@@ -39,40 +40,67 @@ lazy_static! {
     static ref ID: AtomicU64 = AtomicU64::new(0);
 }
 
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, Default)]
 pub struct ChannelManager {
+    inner: Arc<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
     id: u64,
     config: ChannelConfig,
     client_tls_config: Option<ClientTlsConfig>,
     pool: Arc<Pool>,
     channel_recycle_started: Arc<AtomicBool>,
+    cancel: CancellationToken,
 }
 
-impl Default for ChannelManager {
+impl Default for Inner {
     fn default() -> Self {
-        ChannelManager::with_config(ChannelConfig::default())
+        Self::with_config(ChannelConfig::default())
     }
 }
 
-impl ChannelManager {
-    pub fn new() -> Self {
-        Default::default()
+impl Drop for Inner {
+    fn drop(&mut self) {
+        // Cancel the channel recycle task if it is running.
+        if self.channel_recycle_started.load(Ordering::Relaxed) {
+            self.cancel.cancel();
+        }
```

Comment:
Do we need to determine whether the channel recycle task is running?

Reply:
Rust's ownership mechanism ensures that this does not happen.

Reply:
> Rust's ownership mechanism ensures that this does not happen.

Both the `load` and `cas` atomic op have `Relaxed` order so out of order execution could bring surprise.
i.e. if `cas` happen sightly early, then `drop` is called immediately `load` may not read `true`, of course in reality this case might never happen but the potential data race is still here

Reply:
>  of course in reality this case might never happen but the potential data race is still here

Makes sense. While this will never happen, the potential for a data race exists. 

> Do we need to determine whether the channel recycle task is running before cancel?

Then we directly execute `self.cancel.cancel();` in the drop method.

Reply:
Fix in https://github.com/GreptimeTeam/greptimedb/pull/5876/commits/b18831cedffd57a0c4c3c2eac64f1735286f3798